File: path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], k: int) -> int:
        sums = {0: 1}
        ans = 0
        def helper(root, sum_):
            if not root: return
            nonlocal ans
            sum_ += root.val
            ans += sums.get(sum_ - k, 0)
            sums[sum_] = sums.get(sum_, 0) + 1
            helper(root.left, sum_)
            helper(root.right, sum_)
            sums[sum_] -= 1
        helper(root, 0)
        return ans

File: path-sum-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        if(root == nullptr) return 0;
        return get_count(root, targetSum) + pathSum(root->left, targetSum) + pathSum(root->right, targetSum);
    }

    int get_count(TreeNode* node, long long target){
        target -= node->val;
        if(node->left == nullptr && node->right == nullptr){
            return (int)(target == 0);
        }
        int ret = (int)(target == 0);
        if(node->right != nullptr) ret += get_count(node->right, target);
        if(node->left != nullptr) ret += get_count(node->left, target);
        return ret;
    }
};